napari/layers/_shapes_layer/model.py

```diff
@@ -0,0 +1,1521 @@
+        ----------
+        coord : sequence of float
+            Image coordinates to check if any shapes are at.
+
+        Returns
+        ----------
+        shape : int | None
+            Index of shape if any that is at the coordinates. Returns `None`
+            if no shape is found.
+        vertex : int | None
+            Index of vertex if any that is at the coordinates. Returns `None`
+            if no vertex is found.
+        """
+        # Check selected shapes
+        if len(self.selected_shapes) > 0:
+            if self.mode == 'select':
+                # Check if inside vertex of bounding box or rotation handle
+                inds = list(range(0, 8))
+                inds.append(9)
+                box = self._selected_box[inds]
+                distances = abs(box - coord[:2])
+
+                # Get the vertex sizes
+                scene = self.viewer._canvas.scene
+                transform = scene.node_transform(self._node)
+                rescale = transform.map([1, 1])[:2] - transform.map([0, 0])[:2]
+                sizes = self._vertex_size*rescale.mean()/2
+
+                # Check if any matching vertices
+                matches = np.all(distances <= sizes, axis=1).nonzero()
+                if len(matches[0]) > 0:
+                    return self.selected_shapes[0], matches[0][-1]
+            elif (self.mode == 'direct' or self.mode == 'vertex_insert' or
+                  self.mode == 'vertex_remove'):
+                # Check if inside vertex of shape
+                inds = np.isin(self.data._index, self.selected_shapes)
+                vertices = self.data._vertices[inds]
+                distances = abs(vertices - coord[:2])
+
+                # Get the vertex sizes
+                scene = self.viewer._canvas.scene
+                transform = scene.node_transform(self._node)
+                rescale = transform.map([1, 1])[:2] - transform.map([0, 0])[:2]
+                sizes = self._vertex_size*rescale.mean()/2
+
+                # Check if any matching vertices
+                matches = np.all(distances <= sizes, axis=1).nonzero()[0]
+                if len(matches) > 0:
+                    index = inds.nonzero()[0][matches[-1]]
+                    shape = self.data._index[index]
+                    _, idx = np.unique(self.data._index, return_index=True)
+                    return shape, index - idx[shape]
+
+        # Check if mouse inside shape
+        triangles = self.data._mesh_vertices[self.data._mesh_triangles]
+        inside = inside_triangles(triangles - coord[:2])
+        shapes = self.data._mesh_triangles_index[inside]
+
+        if len(shapes) > 0:
+            indices = shapes[:, 0]
+            z_list = self.data._z_order.tolist()
+            order_indices = np.array([z_list.index(m) for m in indices])
+            ordered_shapes = indices[np.argsort(order_indices)]
+            return ordered_shapes[0], None
+        else:
+            return None, None
```

Comment:
This function might also be shortened, though it's not as bad as the first one I mentioned.

Reply:
This one got a little shorter as I did some other refactoring, will leave it for now